### DISCUSSION TITLE
NeighborLoader---I cannot get expected samplers

### DISCUSSION THREAD

**Discussion by ZHAOhngtu** [community/community#138894] [Programming Help]
### Body

The Pytorch program is as below:
import torch
from torch_geometric.datasets import Planetoid

dataset = Planetoid(root='.', name="Pubmed")
data = dataset[0]

# Print information about the dataset
print(f'Dataset: {dataset}')
print('-------------------')
print(f'Number of graphs: {len(dataset)}')
print(f'Number of nodes: {data.x.shape[0]}')
print(f'Number of features: {dataset.num_features}')
print(f'Number of classes: {dataset.num_classes}')

# Print information about the graph
print(f'\nGraph:')
print('------')
print(f'Training nodes: {sum(data.train_mask).item()}')
print(f'Evaluation nodes: {sum(data.val_mask).item()}')
print(f'Test nodes: {sum(data.test_mask).item()}')
print(f'Edges are directed: {data.is_directed()}')
print(f'Graph has isolated nodes: {data.has_isolated_nodes()}')
print(f'Graph has loops: {data.has_self_loops()}')

from torch_geometric.loader import NeighborLoader
from torch_geometric.utils import to_networkx

# Create batches with neighbor sampling
train_loader = NeighborLoader(
    data,
    num_neighbors=[5,10],
    batch_size=16,
    input_nodes=data.train_mask,
)

# Print each subgraph
for i, subgraph in enumerate(train_loader):
    print(f'Subgraph {i}: {subgraph}')

The outputs are 
Number of graphs: 1    
Number of nodes: 19717 
Number of features: 500
Number of classes: 3   

Graph:
------
Training nodes: 60
Evaluation nodes: 500
Test nodes: 1000
Edges are directed: False
Graph has isolated nodes: False
Graph has loops: False
Subgraph 0: Data(x=[16, 500], edge_index=[2, 0], y=[16], train_mask=[16], val_mask=[16], test_mask=[16], n_id=[16], e_id=[0], input_id=[16], batch_size=16)
Subgraph 1: Data(x=[16, 500], edge_index=[2, 0], y=[16], train_mask=[16], val_mask=[16], test_mask=[16], n_id=[16], e_id=[0], input_id=[16], batch_size=16)
Subgraph 2: Data(x=[16, 500], edge_index=[2, 0], y=[16], train_mask=[16], val_mask=[16], test_mask=[16], n_id=[16], e_id=[0], input_id=[16], batch_size=16)
Subgraph 3: Data(x=[12, 500], edge_index=[2, 0], y=[12], train_mask=[12], val_mask=[12], test_mask=[12], n_id=[12], e_id=[0], input_id=[12], batch_size=12)

The subgraphs are incorrect ,How to adjust the program to make them correct ?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by ferid333** [comment]
Hi @ZHAOhngtu. Here’s a checklist and some adjustments you can make to ensure the subgraphs are created correctly:

### Checklist and Adjustments:

1. **Check NeighborLoader Parameters:**
   Ensure the `num_neighbors` parameter in `NeighborLoader` is set according to your graph’s structure. For example, you can try adjusting the `num_neighbors` parameter if the number of neighbors is too high or too low:
   ```python
   train_loader = NeighborLoader(
       data,
       num_neighbors=[10, 10],  # Adjust these numbers if needed
       batch_size=16,
       input_nodes=data.train_mask,
   )
   ```

2. **Verify the Data Masks:**
   Ensure that the `train_mask`, `val_mask`, and `test_mask` are correctly set. You have 60 training nodes, 500 evaluation nodes, and 1000 test nodes. Verify that these masks are properly defined and match your expectations.

3. **Adjust Batch Size:**
   The batch size you’ve set is 16. Ensure that the graph has enough nodes and neighbors to make meaningful batches. You can try adjusting the batch size to see if it affects the subgraphs:
   ```python
   train_loader = NeighborLoader(
       data,
       num_neighbors=[5, 10],
       batch_size=32,  # Try a different batch size
       input_nodes=data.train_mask,
   )
   ```

4. **Check Data Attributes:**
   Verify the attributes of the `data` object. Sometimes, issues arise from incorrectly initialized data. For instance, check that the `data.x`, `data.edge_index`, and other attributes are correctly defined.

5. **Update to Latest Version:**
   Ensure you are using the latest version of PyTorch Geometric, as there may be bug fixes or improvements that could resolve issues with subgraph sampling.

6. **Debug Subgraphs:**
   To understand why the subgraphs may not be correct, you can print additional information:
   ```python
   for i, subgraph in enumerate(train_loader):
       print(f'Subgraph {i}: {subgraph}')
       print(f'Number of nodes in subgraph {i}: {subgraph.num_nodes}')
       print(f'Number of edges in subgraph {i}: {subgraph.num_edges}')
   ```

Here’s how you might adjust your `NeighborLoader` to troubleshoot the issue:

```python
from torch_geometric.loader import NeighborLoader

# Adjust parameters and add debug prints
train_loader = NeighborLoader(
    data,
    num_neighbors=[10, 10],  # Adjust number of neighbors
    batch_size=16,           # Adjust batch size
    input_nodes=data.train_mask,
)

for i, subgraph in enumerate(train_loader):
    print(f'Subgraph {i}: {subgraph}')
    print(f'Number of nodes in subgraph {i}: {subgraph.num_nodes}')
    print(f'Number of edges in subgraph {i}: {subgraph.num_edges}')
    print(f'Node IDs in subgraph {i}: {subgraph.n_id}')
```

**Reply by ZHAOhngtu** [reply]
Hi, ferid333. Thank you very much for the adjustment, Here I present the discrepancy between the output I received and what is shown in a book. I will appreciate it for your running the program and adjust the program to get the book's output. Thank you a lot.
[Discrepancy between the output I received while running the Chapter081.pdf](https://github.com/user-attachments/files/17040644/Discrepancy.between.the.output.I.received.while.running.the.Chapter081.pdf)

**Reply by ferid333** [reply]
Unfortunately, I cannot run the program sorry

**Reply by ZHAOhngtu** [reply]
Hi, ferid333. Thank you very much for reply, Now, I upload the program, the outputs while running it,, and the Pubmed dataset, so that you can run the program. If you spend time to find what caused the discrepancy, that's great. 
[GraphSAGE.zip](https://github.com/user-attachments/files/17060189/GraphSAGE.zip)
